File: figuras/estilo.py
"""
Paleta y configuración compartida de las figuras del capítulo 7.

Todos los scripts de figuras importan de aquí. Nadie teclea un hex suelto
(regla 5 del plan). El teal es el del curso.

Reglas de salida que impone este módulo:
  - svg.fonttype = 'none'  -> el texto queda como texto seleccionable, editable
    en CorelDRAW (regla 3 del plan).
  - guardar() usa metadata={'Date': None} -> sin <dc:date>: el diff de git
    queda limpio al regenerar (regla 4).
  - svg.hashsalt fijo -> los IDs internos del SVG (clips, gradientes) no cambian
    entre corridas, para que el diff sea de verdad limpio.

Además de la paleta trae helpers de rejilla (centro, celda, flecha) para que las
figuras 1, 2 y 3 compartan tamaño de celda y orientación (checklist del plan).
"""
import logging
import re
from pathlib import Path

import matplotlib as mpl
import matplotlib.pyplot as plt
from matplotlib.patches import FancyArrowPatch, Rectangle

# --- Paleta (regla 3 del plan). Único lugar donde vive un hex. ---
TEAL        = "#1a7a8a"   # primario
TEAL_CLARO  = "#2bb5c6"   # secundario / acentos / frentes de wavefront
FONDO_CELDA = "#e0f7fa"   # relleno suave de celda
AMBAR       = "#d98c00"   # alertas / lo que se descarta / inicialización
VERDE       = "#2e7d32"   # el camino óptimo (traceback)
MORADO      = "#7b4fa0"   # cuarto color categórico (ver nota)
GRIS        = "#666666"   # rejilla, ejes
TEXTO       = "#1a1a1a"

# Nota sobre MORADO. Los cuatro colores de arriba alcanzan para casi todo el
# libro, donde lo típico es una serie principal y un contraste. No alcanzan para
# una escala CATEGÓRICA de cinco o seis niveles: las figuras del código genético
# (figuras/sesion03/, en R) colorean los aminoácidos por clase química y
# ácido/básico es justo el contraste que no puede perderse. Vive acá también,
# aunque hoy ninguna figura de Python lo use, porque estilo.py y estilo.R son la
# misma paleta y esa promesa está escrita en el encabezado de los dos archivos.

# Derivados suaves para rellenos. No son hex "sueltos": son versiones claras de
# los tres colores con significado (verde = óptimo, ámbar = init, teal = celda).
VERDE_CLARO = "#d7ebd8"   # relleno de celda en el camino óptimo
AMBAR_CLARO = "#f6e4c3"   # relleno de celda de inicialización
GRIS_CLARO  = "#eef0f1"   # neutro: los ceros interiores de SW ("mar de ceros")

# Tamaño de celda, en unidades de datos. Mismo en las figuras 1, 2 y 3 para que
# la retícula se vea idéntica (argumento pedagógico del plan: es el mismo objeto).
CELDA = 1.0

SVG = Path(__file__).resolve().parent / "svg"

# Stack sans que matplotlib escribe para el texto normal (= el del sitio). Lo
# reusamos en guardar() para reescribir los pocos font-family "sueltos" que deja
# el mathtext y que si no quedarían en la fuente local en vez del stack del libro.
_STACK_SANS = "'Source Sans Pro', 'Segoe UI', 'Roboto', 'Helvetica Neue', 'Arial', sans-serif"
# Un font-family "suelto" es un solo nombre (sin coma) que cierra el valor. Se
# respetan las fuentes de símbolos matemáticos (cm*) y las stacks (llevan coma).
_FONT_SUELTO = re.compile(r"font-family: '(?!cm)[^',]+'(?=[\";])")


def configurar():
    """rcParams comunes. Llamar una vez al principio de cada script."""
    # El texto de las figuras usa el MISMO stack de fuentes que el sitio
    # (assets/css/quarto-lgc.scss): Source Sans Pro para las etiquetas, la
    # monoespaciada del tema para las secuencias. Como svg.fonttype='none', el
    # SVG guarda el nombre de la fuente y el navegador la pinta: al usar el
    # mismo stack que el cuerpo, la figura y el texto combinan —cargue o no
    # Source Sans Pro, porque el fallback es idéntico al del sitio—.
    logging.getLogger("matplotlib.font_manager").setLevel(logging.ERROR)
    mpl.rcParams.update({
        "svg.fonttype": "none",            # texto como texto (regla 3)
        "svg.hashsalt": "bioinfo1-cap07",  # IDs deterministas -> diff limpio
        "font.family": "sans-serif",
        "font.sans-serif": ["Source Sans Pro", "Segoe UI", "Roboto",
                            "Helvetica Neue", "Arial", "sans-serif"],
        "font.monospace": ["SFMono-Regular", "Menlo", "Consolas",
                           "Liberation Mono", "monospace"],
        "font.size": 11,
        "text.color": TEXTO,
        "axes.edgecolor": GRIS,
        "axes.linewidth": 0.8,
        "figure.dpi": 100,
    })


def guardar(fig, nombre, subdir="svg", transparent=False):
    """Guarda el SVG de forma determinista en figuras/<subdir>/<nombre>.svg.

    metadata={'Date': None} quita el <dc:date> que matplotlib mete por defecto;
    sin eso, cada regeneración ensucia el diff de git (regla 4). `subdir` deja
    agrupar por capítulo (p. ej. 'ncbi'); `transparent` para figuras que se
    embeben sin caja de fondo.
    """
    destino = Path(__file__).resolve().parent / subdir
    destino.mkdir(exist_ok=True)
    ruta = destino / f"{nombre}.svg"
    fig.savefig(ruta, format="svg", bbox_inches="tight", metadata={"Date": None},
                transparent=transparent)
    plt.close(fig)
    # El mathtext (ticks de ejes log, $S'$…) escribe la fuente resuelta suelta
    # ('Segoe UI', 'DejaVu Sans'…) en vez del stack. La reescribimos al stack del
    # libro para que TODO el texto combine con el cuerpo, sin tocar símbolos cm*
    # ni la mono (que ya va como stack).
    txt = ruta.read_text(encoding="utf-8")
    txt = _FONT_SUELTO.sub(f"font-family: {_STACK_SANS}", txt)
    ruta.write_text(txt, encoding="utf-8")
    print(f"  escrito figuras/{subdir}/{nombre}.svg")
    return ruta


# --- Helpers de rejilla (compartidos por figs 1, 2, 3, 5, 6) ---

def centro(i, j):
    """Centro (x, y) de la celda (i, j). Fila 0 arriba: y decrece hacia abajo."""
    return (j * CELDA, -i * CELDA)


def celda(ax, i, j, relleno=None, borde=GRIS, lw=0.8, z=1):
    """Dibuja el recuadro de la celda (i, j). Devuelve el Rectangle."""
    x, y = centro(i, j)
    rect = Rectangle((x - CELDA / 2, y - CELDA / 2), CELDA, CELDA,
                     facecolor=relleno if relleno else "none",
                     edgecolor=borde, linewidth=lw, zorder=z)
    ax.add_patch(rect)
    return rect


def flecha(ax, desde_ij, hacia_ij, color=GRIS, lw=1.2, encoge=0.42, alpha=1.0,
           z=3, escala=10):
    """Flecha del centro de `desde_ij` hacia el de `hacia_ij`, encogida en los
    extremos para no encimarse con el número de cada celda. `escala` es el
    tamaño de la punta."""
    x0, y0 = centro(*desde_ij)
    x1, y1 = centro(*hacia_ij)
    dx, dy = x1 - x0, y1 - y0
    d = (dx ** 2 + dy ** 2) ** 0.5
    ux, uy = dx / d, dy / d
    p0 = (x0 + ux * encoge, y0 + uy * encoge)
    p1 = (x1 - ux * encoge, y1 - uy * encoge)
    ax.add_patch(FancyArrowPatch(p0, p1, arrowstyle="-|>", mutation_scale=escala,
                                 color=color, lw=lw, alpha=alpha, zorder=z))


def encabezados(ax, seq_col, seq_fil, cero="", color=TEXTO):
    """Escribe las letras de cada secuencia fuera de la retícula: seq_col arriba,
    seq_fil a la izquierda. `cero` es la etiqueta de la fila/columna 0 (p. ej. '-')."""
    # esquina cero
    if cero:
        xc, yc = centro(0, 0)
        ax.text(xc, yc + CELDA, cero, ha="center", va="center",
                color=color, fontsize=11, fontweight="bold")
        ax.text(xc - CELDA, yc, cero, ha="center", va="center",
                color=color, fontsize=11, fontweight="bold")
    for j, letra in enumerate(seq_col, start=1):
        x, y = centro(0, j)
        ax.text(x, y + CELDA, letra, ha="center", va="center",
                color=color, fontsize=12, fontweight="bold")
    for i, letra in enumerate(seq_fil, start=1):
        x, y = centro(i, 0)
        ax.text(x - CELDA, y, letra, ha="center", va="center",
                color=color, fontsize=12, fontweight="bold")


def marco_limpio(ax, m, n, margen=1.4):
    """Quita ejes y fija límites/relación de aspecto para una retícula (m+1)x(n+1)."""
    ax.set_aspect("equal")
    ax.axis("off")
    ax.set_xlim(-margen, n * CELDA + margen - 0.4)
    ax.set_ylim(-m * CELDA - margen + 0.4, margen)

File: figuras/estilo.R
## Paleta y configuración compartida de las figuras hechas en R (ggplot2 + svglite).
##
## Es el gemelo de estilo.py: MISMA paleta, mismos hex, misma tipografía. Un
## capítulo puede tener figuras en Python y en R y verse igual. Si el tema del
## libro cambia, se cambia en los dos archivos.
##
## Reglas de salida que impone este módulo:
##   - svglite escribe el texto como <text>, no como paths: las etiquetas quedan
##     editables en CorelDRAW. Es la razón de usar svglite y no el device svg()
##     de base R, que vectoriza.
##   - fix_text_size = FALSE: sin atributos textLength. Con textLength puesto,
##     editar una etiqueta en CorelDRAW la estira o la aplasta para respetar el
##     ancho original. Sin él, el texto se comporta como texto normal.
##   - svglite no escribe fecha ni IDs aleatorios (sus ids de clip salen de las
##     coordenadas), así que dos corridas dan bytes idénticos y el diff de git
##     queda limpio al regenerar. Comprobado.
##   - guardar() reescribe el font-family que resuelve svglite (una sola familia,
##     la que haya en la máquina) por el stack completo del libro, para que la
##     figura combine con el cuerpo del sitio aunque Source Sans Pro no esté
##     instalada. Mismo truco que estilo.py.

suppressPackageStartupMessages({
  library(ggplot2)
  library(svglite)
  library(scales)
})

# --- Paleta. Único lugar donde vive un hex (regla compartida con estilo.py). ---
TEAL        <- "#1a7a8a"   # primario
TEAL_CLARO  <- "#2bb5c6"   # secundario / acentos
FONDO_CELDA <- "#e0f7fa"   # relleno suave
AMBAR       <- "#d98c00"   # contraste / referencia / lo que se descarta
VERDE       <- "#2e7d32"   # acento terciario
MORADO      <- "#7b4fa0"   # cuarto color categórico (ver nota)
GRIS        <- "#666666"   # rejilla, ejes, texto secundario
TEXTO       <- "#1a1a1a"

# Nota sobre MORADO. Los cuatro colores de arriba alcanzan para casi todo el
# libro, donde lo típico es una serie principal y un contraste. No alcanzan
# para una escala CATEGÓRICA de cinco o seis niveles: las figuras del código
# genético colorean los aminoácidos por clase química (hidrofóbico, polar,
# ácido, básico, especial) y ácido/básico son justo el contraste que no puede
# perderse. Con TEAL y TEAL_CLARO esos dos quedaban a un paso de distancia en
# el mismo tono. De ahí el quinto tono.
# Está en estilo.py también, para no romper el pacto de que las dos paletas son
# la misma aunque hoy ninguna figura de Python lo use.

# El mismo stack que assets/css/quarto-lgc.scss. Va con comillas DOBLES porque
# svglite delimita el atributo style con comillas simples.
STACK_SANS <- paste0('"Source Sans Pro", "Segoe UI", "Roboto", ',
                     '"Helvetica Neue", "Arial", sans-serif')

# Stack monoespaciado, para las figuras de esquema donde el texto ES código
# (anatomía de un FASTA, un pipeline) y tiene que caer en una retícula de ancho
# fijo. Mismo criterio que STACK_SANS: se escribe el stack completo, no la
# familia que resolvió esta máquina.
STACK_MONO <- paste0('"Source Code Pro", "Consolas", "DejaVu Sans Mono", ',
                     '"Courier New", monospace')

# Preferencias de familia, en orden. Se usa la primera instalada para que R no
# tire warnings de fuente faltante; el SVG final lleva el stack completo de
# todos modos, así que en el navegador manda el stack, no esta elección.
.FAMILIAS      <- c("Source Sans Pro", "Segoe UI", "Roboto", "Helvetica Neue", "Arial")
.FAMILIAS_MONO <- c("Source Code Pro", "Consolas", "DejaVu Sans Mono",
                    "Courier New", "Liberation Mono", "Menlo", "monospace")


familia_base <- function() {
  instaladas <- unique(systemfonts::system_fonts()$family)
  hit <- .FAMILIAS[.FAMILIAS %in% instaladas]
  if (length(hit)) hit[1] else ""
}


#' Gemela de familia_base() para texto que es código.
familia_mono <- function() {
  instaladas <- unique(systemfonts::system_fonts()$family)
  hit <- .FAMILIAS_MONO[.FAMILIAS_MONO %in% instaladas]
  if (length(hit)) hit[1] else "mono"
}


#' Tema del libro. Equivale a los rcParams de estilo.py::configurar().
tema_lgc <- function(base_size = 11) {
  theme_minimal(base_size = base_size, base_family = familia_base()) +
    theme(
      text             = element_text(colour = TEXTO),
      axis.title       = element_text(size = rel(0.96), colour = TEXTO),
      axis.text        = element_text(size = rel(0.86), colour = TEXTO),
      axis.line        = element_line(colour = GRIS, linewidth = 0.35),
      axis.ticks       = element_line(colour = GRIS, linewidth = 0.35),
      axis.ticks.length = unit(2.5, "pt"),
      panel.grid.major = element_line(colour = alpha(GRIS, 0.16), linewidth = 0.35),
      panel.grid.minor = element_blank(),
      panel.background = element_blank(),
      plot.background  = element_blank(),
      legend.background = element_blank(),
      legend.key       = element_blank(),
      legend.title     = element_blank(),
      legend.text      = element_text(size = rel(0.86)),
      # Nota de fuente: gris, chica, en cursiva y pegada al borde izquierdo de
      # la figura (no del panel), como la ponía estilo.py.
      plot.caption     = element_text(size = rel(0.73), colour = GRIS,
                                      face = "italic", hjust = 0,
                                      margin = margin(t = 8)),
      plot.caption.position = "plot",
      plot.title       = element_blank(),   # el caption vive en el .qmd
      plot.margin      = margin(4, 6, 4, 4)
    )
}


#' Guarda el SVG de forma determinista en figuras/<subdir>/<nombre>.svg.
#'
#' @param p       objeto ggplot
#' @param nombre  sin extensión
#' @param subdir  subcarpeta dentro de figuras/ (p. ej. "sesion01")
#' @param ancho,alto  pulgadas
#' @param transparent  sin caja de fondo, para embeber en el sitio
guardar <- function(p, nombre, subdir = "svg", ancho = 6.3, alto = 3.9,
                    transparent = TRUE) {
  destino <- file.path(DIR_FIGURAS, subdir)
  dir.create(destino, showWarnings = FALSE, recursive = TRUE)
  ruta <- file.path(destino, paste0(nombre, ".svg"))

  svglite(ruta, width = ancho, height = alto,
          bg = if (transparent) "transparent" else "white",
          fix_text_size = FALSE)
  print(p)
  invisible(grDevices::dev.off())

  # svglite resuelve UNA familia ("Arial" si no hay Source Sans Pro). Se cambia
  # por el stack del libro para que el navegador use la misma cascada que el
  # cuerpo del texto. No se tocan otras propiedades del style.
  txt <- readLines(ruta, warn = FALSE, encoding = "UTF-8")

  # Las monoespaciadas se marcan ANTES para que el reemplazo de abajo no se las
  # lleve por delante. Sin esto, una figura de esquema calcula su retícula con
  # los avances de una mono y el navegador la rellena con glifos
  # proporcionales: las columnas dejan de cuadrar y las letras se encinan.
  # (Pasó, y no se ve en R: sólo aparece después del post-proceso.)
  for (fm in .FAMILIAS_MONO) {
    txt <- gsub(sprintf('font-family: "%s";', fm), "font-family: @@MONO@@;",
                txt, fixed = TRUE)
  }
  txt <- gsub('font-family: "[^"]+";', paste0("font-family: ", STACK_SANS, ";"),
              txt, perl = TRUE)
  txt <- gsub("font-family: @@MONO@@;", paste0("font-family: ", STACK_MONO, ";"),
              txt, fixed = TRUE)
  con <- file(ruta, open = "wb")
  writeLines(txt, con, useBytes = TRUE)
  close(con)

  message(sprintf("  escrito figuras/%s/%s.svg", subdir, nombre))
  invisible(ruta)
}


# --- Ubicación de figuras/ -------------------------------------------------
# R no tiene __file__. Se busca el `ofile` que source() deja en los frames de
# llamada; si estilo.R se corre directo con Rscript, se usa --file=.
.ubicar_figuras <- function() {
  # De ADENTRO hacia afuera: el `ofile` más cercano es el de este archivo. Si se
  # recorriera al revés y alguien sourcea un script de figura desde otro script
  # (p. ej. un "regenerar todo"), el primer ofile sería el del script de afuera
  # y figuras/ quedaría mal resuelto: los SVG terminarían en sesion01/sesion01/.
  for (i in rev(seq_len(sys.nframe()))) {
    of <- sys.frame(i)$ofile
    if (!is.null(of)) return(dirname(normalizePath(of, winslash = "/")))
  }
  a <- grep("^--file=", commandArgs(trailingOnly = FALSE), value = TRUE)
  if (length(a)) {
    return(dirname(normalizePath(sub("^--file=", "", a[1]), winslash = "/")))
  }
  normalizePath(getwd(), winslash = "/")
}

DIR_FIGURAS <- .ubicar_figuras()


# --- Formateadores de eje compartidos --------------------------------------

# Un número sin notación científica y sin decimales de más. OJO: format() es
# vectorizado y elige UN formato común para todo el vector, así que sobre un
# eje log convierte el 1000 en "1e+03". Por eso los formateadores de abajo
# trabajan elemento por elemento con formatC.
.num <- function(x, dig = 0) formatC(x, format = "f", digits = dig)

# Fábrica de formateadores por escala: devuelve una función de eje.
.fmt_escala <- function(escalas, sep = "", prefijo = "") {
  function(v) {
    vapply(v, function(x) {
      if (is.na(x)) return(NA_character_)
      for (e in escalas) {
        u <- as.numeric(e[[1]])
        if (abs(x) >= u) return(paste0(prefijo, .num(x / u), sep, e[[2]]))
      }
      paste0(prefijo, .num(x))
    }, character(1))
  }
}

#' Dinero corto y legible: $100, $1K, $100K, $1M, $100M.
fmt_dolar <- .fmt_escala(list(list(1e6, "M"), list(1e3, "K")), prefijo = "$")

#' Bases en Kb / Mb / Gb / Tb.
fmt_bases <- .fmt_escala(list(list(1e12, "Tb"), list(1e9, "Gb"),
                              list(1e6, "Mb"), list(1e3, "Kb")), sep = " ")

#' Conteos en K / M / G.
fmt_conteo <- .fmt_escala(list(list(1e9, "G"), list(1e6, "M"), list(1e3, "K")),
                          sep = " ")

#' Año decimal a partir de (año, mes), tomando el punto medio del mes.
a_fraccion <- function(anio, mes) anio + (mes - 0.5) / 12
